network/v1alpha1/types_dnsnameresolver.go

```diff
@@ -0,0 +1,137 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +openshift:compatibility-gen:level=4
+
+// DNSNameResolver stores the DNS name resolution information of a DNS name. It can be enabled by the TechPreviewNoUpgrade feature set.
+// It can also be enabled by the feature gate DNSNameResolver when using CustomNoUpgrade feature set.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+type DNSNameResolver struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the DNSNameResolver.
+	// +kubebuilder:validation:Required
+	Spec DNSNameResolverSpec `json:"spec"`
+	// status is the most recently observed status of the DNSNameResolver.
+	// +optional
+	Status DNSNameResolverStatus `json:"status,omitempty"`
+}
+
+// DNSName is used for validation of a DNS name.
+// +kubebuilder:validation:Pattern=`^(\*\.)?([a-z0-9]([-a-z0-9]{0,61}[a-z0-9])?\.){2,}$`
```

Comment:
I'm wondering if we could break down this validation at all into CEL and then use that to enforce the different rules and apply better error messages. Will have a think about this.

Reply:
@JoelSpeed The `dnsName` field is used inside the `resolvedNames` field. We will probably hit the `cost exceeds budget` issue with CEL validations that we faced while trying to add the same for the `ip` field: https://github.com/openshift/api/pull/1524#discussion_r1302583351

Reply:
I'm also a little worried that this might cause issues for DNS names that are accepted by the library-go validation for Route but not accepted here.  In a Route, we had to allow for a one-label host name:  https://github.com/openshift/library-go/blob/master/pkg/route/routeapihelpers/routeapihelpers.go#L63

Reply:
This should not be a problem, as while creating the CR the OVN-K master will correctly convert a DNS name into this format. Even the EgressFirewall DNS name accepts DNS names without trailing periods, however while creating the corresponding DNSNameResolver CR, OVN-K master will add the trailing period. The more strict regex is added as we want to eliminate the chances of creating more than one CR for the same DNS name.

Reply:
Ahh yes, we can't do this because `resolvedAddresses` is an unbounded list. I don't think given previous conversation that we can bound resolvedAddresses, so nothing to do here